### PR TITLE
net/portmapper: fix UPnP probing

### PR DESCRIPTION
**Pull request by bradfitz**
Prior to Tailscale 1.12 it detected UPnP on any port.
Starting with Tailscale 1.11.x, it stopped detecting UPnP on all ports.

Then use the Location header to find the rootDesc.xml, rather than hard-coding the URL + port.

And remove the three goroutines doing redundant & racing rootDesc.xml fetches.

And use netns so it works while using exit nodes.

Works against pfSense now.

Updates #2109

### CODE REVIEW

**Comment by bradfitz** [comment]
I'm now getting, against pfSense:

```
root@tailtest:~# ./tailscaled debug -portmap
2021/08/02 22:07:16 portmapper: unexpected PMP probe response: {OpCode:128 ResultCode:3 SecondsSinceEpoch:8671 MappingValidSeconds:0 InternalPort:0 ExternalPort:0 PublicAddr:0.0.0.0}
2021/08/02 22:07:16 portmapper: Got PCP response: epoch: 8671
2021/08/02 22:07:16 UPnP reply from 192.168.1.1:2189: HTTP/1.1 200 OK
CACHE-CONTROL: max-age=120
ST: urn:schemas-upnp-org:device:InternetGatewayDevice:1
USN: uuid:bee7052b-49e8-3597-b545-55a1e38ac11::urn:schemas-upnp-org:device:InternetGatewayDevice:1
EXT:
SERVER: FreeBSD/12.2-STABLE UPnP/1.1 MiniUPnPd/2.2.1
LOCATION: http://192.168.1.1:2189/rootDesc.xml
OPT: "http://schemas.upnp.org/upnp/1/0/"; ns=01
01-NLS: 1627958564
BOOTID.UPNP.ORG: 1627958564
CONFIGID.UPNP.ORG: 1337

2021/08/02 22:07:16 Probe: {PCP:true PMP:false UPnP:true}
2021/08/02 22:07:16 no mapping
2021/08/02 22:07:16 Getting root desc: http://192.168.1.1:2189/rootDesc.xml
2021/08/02 22:07:16 XXX meth1: [0xc000315200]
2021/08/02 22:07:16 XXX client = &{{0xc000336000 0xc000212000 http://192.168.1.1:2189/rootDesc.xml Service ID urn:upnp-org:serviceId:WANIPConn1 : urn:schemas-upnp-org:service:WANIPConnection:1}}
2021/08/02 22:07:16 XXX meth2: error requesting root device details from "http://192.168.1.1:2189/rootDesc.xml": Get "http://192.168.1.1:2189/rootDesc.xml": context canceled
2021/08/02 22:07:16 XXX meth3: error requesting root device details from "http://192.168.1.1:2189/rootDesc.xml": Get "http://192.168.1.1:2189/rootDesc.xml": context canceled
2021/08/02 22:07:16 XXX addAny = SOAP fault: 
```

File: net/portmapper/upnp.go [hunk 1]
```diff
@@ -130,40 +132,70 @@ func addAnyPortMapping(
...
-	ctx, cancel := context.WithTimeout(ctx, 250*time.Millisecond)
-	defer cancel()
-	// Attempt to connect over the multiple available connection types concurrently,
-	// returning the fastest.
 
-	// TODO(jknodt): this url seems super brittle? maybe discovery is better but this is faster
-	u, err := url.Parse(fmt.Sprintf("http://%s:5000/rootDesc.xml", gw))
+	var port uint16
+	if uPnPRootIPP.IsZero() {
+		port = 5000
```
Comment (JulianKnodt) [inline net/portmapper/upnp.go:150]: Is there anyway to unset root IP if a network change is detected?

File: net/portmapper/upnp.go [hunk 2]
```diff
@@ -130,40 +132,70 @@ func addAnyPortMapping(
...
+	} else {
+		port = uPnPRootIPP.Port()
+		if uPnPRootIPP.IP() != gw {
+			return nil, fmt.Errorf("UPnP discovered root %v != gateway %v; ignoring UPnP",
+				uPnPRootIPP, gw)
+		}
+	}
+
+	rootDescURL := fmt.Sprintf("http://%s/rootDesc.xml", net.JoinHostPort(gw.String(), fmt.Sprint(port)))
```
Comment (JulianKnodt) [inline net/portmapper/upnp.go:159]: One other thing I noticed is that I think the caps on `rootDesc` is not capitalized the same way across all UPnP platforms, just something to note for the future

Comment (bradfitz) [reply net/portmapper/upnp.go]: Fixed.

File: net/portmapper/disabled_stubs.go [hunk 3]
```diff
@@ -15,10 +15,6 @@ import (
 
 type upnpClient interface{}
 
-func getUPnPClient(ctx context.Context, gw netaddr.IP) (upnpClient, error) {
-	return nil, nil
-}
```
Comment (JulianKnodt) [inline net/portmapper/disabled_stubs.go:20]: Is it fine to delete this for iOS now? How will it handle having no stub here?

File: net/portmapper/portmapper.go [hunk 4]
```diff
@@ -612,6 +604,19 @@ func (c *Client) Probe(ctx context.Context) (res ProbeResult, err error) {
 		}
 		port := addr.(*net.UDPAddr).Port
 		switch port {
+		case upnpPort:
+			if mem.Contains(mem.B(buf[:n]), mem.S(":InternetGatewayDevice:")) {
+				meta, err := parseUPnPDiscoResponse(buf[:n])
+				if err != nil {
+					c.logf("unrecognized UPnP discovery response; ignoring")
+				}
+				// log.Printf("UPnP reply %+v, %q", meta, buf[:n])
```
Comment (JulianKnodt) [inline net/portmapper/portmapper.go:613]: Probably want to delete this before merging

File: net/portmapper/portmapper.go [hunk 5]
```diff
@@ -62,8 +64,11 @@ type Client struct {
 	pmpPubIPTime time.Time  // time pmpPubIP last verified
 	pmpLastEpoch uint32
 
-	pcpSawTime  time.Time // time we last saw PCP was available
-	uPnPSawTime time.Time // time we last saw UPnP was available
+	pcpSawTime time.Time // time we last saw PCP was available
+
+	uPnPSawTime    time.Time         // time we last saw UPnP was available
+	uPnPMeta       uPnPDiscoResponse // Location header from UPnP UDP discovery response
+	uPnPHTTPClient *http.Client      // nil until needed
```
Comment (JulianKnodt) [inline net/portmapper/portmapper.go:71]: I think it's worth adding a note what this field is for

File: net/portmapper/portmapper.go [hunk 6]
```diff
@@ -724,3 +729,13 @@ func parsePCPResponse(b []byte) (res pcpResponse, ok bool) {
 }
 
 var pmpReqExternalAddrPacket = []byte{0, 0} // version 0, opcode 0 = "Public address request"
+
+const (
+	upnpPort = 1900
```
Comment (JulianKnodt) [inline net/portmapper/portmapper.go:734]: I'm not sure if this will properly detect UPnP on my router, but I'll check it when this gets merged

**Review by JulianKnodt — APPROVED** [review]

File: net/portmapper/portmapper.go (hunk 6 above)
Comment (bradfitz) [reply net/portmapper/portmapper.go]: This is only the UDP discovery port number, which is a fixed number. I added a comment.

File: net/portmapper/upnp.go (hunk 1 above)
Comment (bradfitz) [reply net/portmapper/upnp.go]: Cleared in invalidateMappingsLocked now.

File: net/portmapper/portmapper.go (hunk 5 above)
Comment (bradfitz) [reply net/portmapper/portmapper.go]: Done.

File: net/portmapper/portmapper.go (hunk 4 above)
Comment (bradfitz) [reply net/portmapper/portmapper.go]: Updated.

File: net/portmapper/upnp.go [hunk 7]
```diff
@@ -130,51 +141,85 @@ func addAnyPortMapping(
...
 		return nil, nil
 	}
-	ctx, cancel := context.WithTimeout(ctx, 250*time.Millisecond)
+
+	if meta.Location == "" {
+		return nil, nil
+	}
+
+	if VerboseLogs {
+		log.Printf("fetching %v", meta.Location)
```
Comment (crawshaw) [inline net/portmapper/upnp.go:165]: logf

File: net/portmapper/upnp.go [hunk 8]
```diff
@@ -130,51 +141,85 @@ func addAnyPortMapping(
...
+	ipp, err := netaddr.ParseIPPort(u.Host)
+	if err != nil {
+		return nil, fmt.Errorf("unexpected host %q in %q", u.Host, meta.Location)
+	}
+	if ipp.IP() != gw {
+		return nil, fmt.Errorf("UPnP discovered root %q does not match gateway IP %v; ignoring UPnP",
+			meta.Location, gw)
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, 500*time.Millisecond)
```
Comment (crawshaw) [inline net/portmapper/upnp.go:181]: any intuition for this timeout? add a comment if there is one?

**Review by crawshaw — APPROVED** [review]
woo tests

File: net/portmapper/upnp.go (hunk 8 above)
Comment (bradfitz) [reply net/portmapper/upnp.go:181]: Done. And doubled it:
```go
        // We're fetching a smallish XML document over plain HTTP                                                                                                                          
        // across the local LAN, without using DNS.  There should be                                                                                                                       
        // very few round trips and low latency, so one second is a                                                                                                                        
        // long time.                                                                                                                                                                      
        ctx, cancel := context.WithTimeout(ctx, time.Second)
        defer cancel()
```

**Pull request merged**